pkg/compactor/block_upload.go

```diff
@@ -262,29 +285,73 @@ func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Re
 	w.WriteHeader(http.StatusOK)
 }
 
-func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) error {
+func (c *MultitenantCompactor) validateAndCompleteBlockUpload(logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta *metadata.Meta) {
 	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
 
-	// Upload meta file so block is considered complete
+	ch := make(chan struct{})
+	var wg sync.WaitGroup
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start a go routine that updates the validation file's timestamp every heartbeat interval
+	wg.Add(1)
+	go c.periodicValidationUpdater(ctx, logger, blockID, userBkt, ch, &wg, cancel)
+
+	if err := c.validateBlock(ctx, blockID, userBkt, meta); err != nil {
+		level.Error(logger).Log("msg", "error while validating block", "err", err)
+		close(ch)
+		wg.Wait()
+		if !errors.Is(err, context.Canceled) {
+			err := c.uploadValidationWithError(ctx, blockID, userBkt, err.Error())
```

Comment:
Here we could just use fresh `context.Background` to upload validation file with error, and not bother checking if `ctx` is still usable.

It would simplify the code a bit, although technically if context was canceled, we know that `c.periodicValidationUpdater` failed... but it's hard to see that in the code. I'd say simpler code wins here.

Reply:
Agreed, removed these context checks.